**Задание 2: Автоматическое дифференцирование (25 баллов)** \
**2.1 Простые вычисления с градиентами (8 баллов)**

In [ ]:
import torch

In [5]:
x = torch.rand(2, requires_grad=True)
y = torch.rand(2, requires_grad=True)
z = torch.rand(2, requires_grad=True)
f = (x**2 + y**2 + z**2 + 2*x*y*z).sum()
f.backward()
print(f'x {x}')
print(f'y {y}')
print(f'z {z}')
print(x.grad)
print(y.grad)
print(z.grad)

x tensor([0.9407, 0.4009], requires_grad=True)
y tensor([0.4875, 0.9949], requires_grad=True)
z tensor([0.1093, 0.0363], requires_grad=True)
tensor([1.9879, 0.8740])
tensor([1.1805, 2.0190])
tensor([1.1357, 0.8704])


**Проверка результата** \
f(x,y,z) = x^2 + y^2 + z^2 + 2*x*y*z \
df/dx = 2x + 0 + 0 + 2yz = 2x + 2yz \
df/dy = 2y + 2xz \
df/dz = 2z + 2xy \
Подставим: \
**df/dx** = 2 * [0.9407, 0.4009] + 2 * [0.4875, 0.9949] * [0.1093, 0.0363] = [1.8814, 0.8018] + 2 * [0.0533, 0.0361] = \
 = [1.8814, 0.8018] + [0.1066, 0.0722] = **[1.9880, 0.8740]** \
**df/dy** = 2 * [0.4875, 0.9949] + 2 * [0.9407, 0.4009] * [0.1093, 0.0363] = [0.9750, 1.9898] + 2 * [0.1029, 0.0146] = \
 = [0.9750, 1.9898] + [0.2058, 0.0292] = **[1.1808, 2.0190]** \
**df/dz** = 2 * [0.1093, 0.0363] + 2 * [1.9879, 0.8740] * [1.1805, 2.0190] = [0.2186, 0.0726] + 2 * [0.4585, 0.3989] = \
 = [0.2186, 0.0726] + [0.9170, 0.7978] = **[1.1356, 0.8704]** \
Результаты одинаковы.

**2.2 Градиент функции потерь (9 баллов)**

In [12]:
x = torch.randint(1, 10, (2,), dtype=torch.float32, requires_grad=True) #тензор x из случайных целых чисел от 1 до 9, размерностью (2,)
y_true = torch.rand(2, requires_grad=True) #тензор y_true со случайными значениями от 0 до 1, размерностью (2,)
w = torch.tensor(1.0, requires_grad=True)
b = torch.tensor(2.0, requires_grad=True)
y_pred = w * x + b
# MSE = (1/n) * Σ(y_pred - y_true)^2
mse = ((y_pred - y_true) ** 2).sum()
mse.backward()
print(w.grad)
print(b.grad)

tensor(112.4291)
tensor(22.9556)


**2.3 Цепное правило (8 баллов)**

In [15]:
x = torch.tensor(1.0, requires_grad=True) #создаем тензор
f = torch.sin(x**2 + 1) #высчитываем функцию
f.backward()
print(x.grad)
x.grad.zero_() #очищаем тензор
f = torch.sin(x**2 + 1) #заново высчитываем функцию
grad = torch.autograd.grad(f, x)
print(grad)

tensor(-0.8323)
(tensor(-0.8323),)


**Ручная проверка** \
**df/dx** = 2x * cos(x^2+1) = 1*cos(2) = 2 * (-0.4161) = -0.8322